### DISCUSSION TITLE
How to get host memory from a host function?

### DISCUSSION THREAD

**Discussion by orsinium** [wasmi-labs/wasmi#936] [Q&A] [answered]
I can define a host function that uses **module-defined** memory like this:

```rust
linker.func_wrap("env", "hline", |mut caller: C, x: i32, y: i32, len: u32| {
    let mem = match caller.get_export("memory") {
        Some(wasmi::Extern::Memory(mem)) => mem,
        _ => panic!("memory not found"),
    };
    let (data, bridge) = mem.data_and_store_mut(&mut caller);
    let frame_buf = FrameBuf::from_memory(data);
    bridge.wasm4_hline(frame_buf, x, y, len)
})?;
```

However, if I define a memory on the host side:

```rust
let mem_type = wasmi::MemoryType::new(1, Some(1)).unwrap();
let mem = wasmi::Memory::new(&mut store, mem_type).unwrap();
linker.define("env", "memory", mem).unwrap();
```

I can't simply use `mem` after that because `linker.define` takes ownership of it. I need to get is somehow back from the store. Outside of the host-defined functions, I can do something like this:

```rust
let memory = match linker.get(&mut store, "env", "memory") {
    Some(wasmi::Extern::Memory(memory)) => memory,
    _ => panic!("memory not found"),
};
```

However, if I try to do the same inside of a `linker.func_wrap` closure, I get a compilation error:

```
borrowed data escapes outside of function
`linker` escapes the function body here
```

Is there a way to get access to the memory inside the linker from a `wasmi::Func`? I don't see how to keep its ownership after calling `linker.define` and I don't see any method for that in the `wasmi::Caller`.

**Comment by Robbepop** [comment, marked as answer]
Wasm types such as `Memory`, `Table`, `Global` and `Func` are deliberately `Copy` and `Clone` types and thus cheap to clone. This works since they are just fancy, typed indices into some tables into the `Store`.

Thus you can simply copy the `mem` before passing it to the `Linker` in one of your examples and even put `mem` into the `HostState` to access it via host functions amongst others.
